### PR TITLE
`isDirty` flag

### PR DESCRIPTION
**Pull request by weili-govtech**
**Changes**
- added `isDirty` flag which can be referenced from frontend engine ref
- fixed setting of dirty states on mount / prefilling / reset for
  - contact-field
  - date-field
  - image-upload
  - location-field
- added test cases for frontend engine and the 5 components
- test cases for the remaining components will be added in another PR

### CODE REVIEW

File: src/components/fields/contact-field/contact-field.tsx [hunk 1]
```diff
@@ -118,9 +120,11 @@ export const ContactField = (props: IGenericFieldProps<IContactFieldSchema>) =>
 			const schemaPrefix = getPrefixFromCountry(countryValue || "Singapore");
 			if (!fixedCountry || (fixedCountry && prefixWithoutPlus === schemaPrefix)) {
 				const fieldPrefix = prefix || schemaPrefix;
-				handleChange({ countryCode: fieldPrefix, number });
+				const phoneNumberWithPrefix = parseAndApplyNumber({ countryCode: fieldPrefix, number });
+				setValue(id, phoneNumberWithPrefix, { shouldDirty: false });
```
Comment (qroll) [inline src/components/fields/contact-field/contact-field.tsx:124]: default `shouldDirty` is false so the second args can be omitted

**Review by qroll — COMMENTED** [review]
just some minor comments, played around in storybook and seems fine

an observation I had is that `ref.current.isDirty` is not immediately reflected in `onChange`, but this should not be a problem for our service use case as the value will be checked outside of `onChange`

File: src/components/frontend-engine/frontend-engine.tsx [hunk 2]
```diff
@@ -169,6 +170,8 @@ const FrontendEngineInner = forwardRef<IFrontendEngineRef, IFrontendEngineProps>
 	}, [checkIsFormValid, onChange, watch, formValidationConfig]);
 
 	useEffect(() => {
+		// console.log("*** rhf isDirty", formState.isDirty);
+		// console.log("*** rhf touchedFields", formState.touchedFields);
```
[the remark this answers is not in the document]
Comment (weili-govtech) [reply src/components/frontend-engine/frontend-engine.tsx]: have removed

File: src/components/fields/image-upload/image-manager/image-manager.ts [hunk 3]
```diff
@@ -137,17 +138,19 @@ export const ImageManager = (props: IProps) => {
...
-				value: images
-					.filter(({ status }) => status === EImageStatus.UPLOADED)
-					.map(({ dataURL, drawingDataURL, name, uploadResponse }) => ({
-						fileName: name,
-						dataURL: drawingDataURL || dataURL,
-						uploadResponse,
-					})),
-			},
-		});
+		const shouldDirty = images.filter(({ addedFrom }) => addedFrom !== "schema").length > 0;
```
[the remark this answers is not in the document]
Comment (weili-govtech) [reply src/components/fields/image-upload/image-manager/image-manager.ts]: I didn't manage to find a reliable way to determine a genuine delete attempt vs reset so I implemented a new status:
- new `TO_DELETE` image status
- when user initiates to delete an image, the image status will be updated to `TO_DELETE`
- the image manager will delete it in useEffect
- image manager sets the isDirty state based on this status as well

File: src/components/fields/contact-field/contact-field.tsx (hunk 1 above)
Comment (weili-govtech) [reply src/components/fields/contact-field/contact-field.tsx]: have removed various instances of it

**Comment by weili-govtech** [comment]
> an observation I had is that ref.current.isDirty is not immediately reflected in onChange, but this should not be a problem for our service use case as the value will be checked outside of onChange

Yeah I think it's fine, if anyone wants to get the updated `isDirty` flag, a `setTimeout` will work too.

**Review by qroll — APPROVED** [review]

**Pull request merged**